lib/rucio/core/rse.py

```diff
@@ -168,6 +171,19 @@ def del_rse(rse, session=None):
     del_rse_attribute(rse=rse, key=rse, session=session)
 
 
+@read_session
+def rse_is_empty(rse, session=None):
+    """
+    Check if no file replica belongs to the RSE.
+
+    :param rse: the rse name.
+    :param session: the database session in use.
+    """
+
+    rse_id = get_rse(rse).id
+    return session.query(models.RSEFileAssociation).filter_by(rse_id=rse_id).count() == 0
```

Comment:
This can take a very long time. The table is not indexed by `rse_id` and it will go to full table scan. We might need to set an index. Other possibility : use the replicas dumps.

Reply:
another index on replicas is maybe not so good, it will blow up the tablespace. what about using collection replicas? that would be much smaller and faster. for the result from collection replicas an additional deep check to replicas can be done afterwards?

Reply:
or even rse counters?

Reply:
ah yes, rse counters makes more sense. It already has the information about the rse usage. I guess there is some delay in updating it but I don't think this is a big problem?

Reply:
that's not really a problem. the more problematic part is when the abacus fails and things are at zero and then this function would believe it's empty when in fact it's not.

Reply:
I guess collection replicas can not be used also, because there can be files on the RSE that are not attached to a collection, if I understood it correctly.

And what is a replica dump actually?

Reply:
Replica dump is not a good idea I think. It seems wrong to introduce this on server nodes.
I think rse_counters is a good idea. They are sometimes out of sync, but deleting an rse is a rare occassion. If the deletion fails because of the counters being out of sync, we just have to re-sync the counters.
👍 for counters

Reply:
Btw. Just for the record: I tried to run this query on a deleted_rse and it takes about 50m to finish. On the ATLAS instance (10m timeout) it would never finish 😄